### PR TITLE
ENH: umath: ensure ufuncs are well-defined with memory overlapping inputs

### LINKED ISSUE #1683: Surprising results from in-place operations involving views (Trac #1085)
_Original ticket http://projects.scipy.org/numpy/ticket/1085 on 2009-04-15 by trac user tillmann, assigned to @charris._

> > > import numpy as np
> > > x=np.array([[1,2],[3,4]])
> > > print x
> > > [[1 2]
> > >  [3 4]]
> > > x+=x.T
> > > print x
> > > [[2 5]
> > >  [8 8]]

when writing code on such a high level of abstraction this behavior is not expected and imho wrong. At least a warning should be issued or better the in place operations should fall back to creating copies if the results are not correct otherwise.

### PR DESCRIPTION
**Pull request by pv**
Make ufuncs and ufunc operations to make temporary copies if input and output operands overlap in memory. Currently (cf. gh-6272, gh-1683) output from such operations is undefined. With this PR, the result is the same as if the operands were non-overlapping.

Fixes gh-1683

Adds new flags to NpyIter, and one new API function is added to deal with MapIter.

This PR does not fix up overlap handling in array assignment, that needs to be done later (and is slightly more hairy --- some common overlapping cases are in fact already handled). Unlike adding overlap-if-copy flag to NpyIter, it appears a similar flag would not be as useful in MapIter, as there are many special-cased code paths, and MapIter is not always made aware of all arrays operated on.

- [x] reread whole patch
- [x] re-check test coverage, and whether some of the tests are unnecessary

### CODE REVIEW

File: numpy/core/src/private/lowlevel_strided_loops.h [hunk 1]
```diff
@@ -673,6 +691,60 @@ npy_bswap8_unaligned(char * x)
                                               NPY_ARRAY_F_CONTIGUOUS)) \
                         )
 
+#define PyArray_TRIVIAL_PAIR_ITERATION_STRIDE(size, arr) ( \
+                        size == 1 ? 0 : ((PyArray_NDIM(arr) == 1) ? \
+                                          PyArray_STRIDE(arr, 0) : \
+                                          PyArray_ITEMSIZE(arr)))
+
+static NPY_INLINE int
+PyArray_EQUIVALENTLY_ITERABLE_OVERLAP_OK(PyArrayObject *arr1, PyArrayObject *arr2,
```
Comment (charris) [inline numpy/core/src/private/lowlevel_strided_loops.h]: Seems rather large for inline ;)

File: numpy/core/tests/test_mem_overlap.py [hunk 2]
```diff
@@ -4,9 +4,11 @@
 import itertools
 
 import numpy as np
-from numpy.testing import run_module_suite, assert_, assert_raises, assert_equal
+from numpy.testing import (run_module_suite, assert_, assert_raises, assert_equal, assert_array_equal,
```
Comment (charris) [inline numpy/core/tests/test_mem_overlap.py]: Line too long ?

File: numpy/core/code_generators/numpy_api.py [hunk 3]
```diff
@@ -344,6 +344,8 @@
     # End 1.9 API
     'PyArray_CheckAnyScalarExact':          (300, NonNull(1)),
     # End 1.10 API
+    # End 1.11 API
```
Comment (charris) [inline numpy/core/code_generators/numpy_api.py]: Should update this now that 1.12 is branched. The cversion and C-API number also need updating, see documentation in `doc/HOWTO_RELEASE.rst.txt`.

**Comment by charris** [comment]
Just starting to look at this.

File: numpy/core/src/umath/ufunc_object.c [hunk 4]
```diff
@@ -1777,22 +1781,30 @@ execute_fancy_ufunc_loop(PyUFuncObject *ufunc,
...
-                            arr_prep[i], arr_prep_args, i) < 0) {
+        if (PyArray_BYTES(op_tmp) != PyArray_BYTES(op_it[i])) {
+            PyErr_SetString(PyExc_ValueError,
+                        "The __array_prepare__ functions modified the data "
+                        "pointer addresses in an invalid fashion");
+            Py_DECREF(op_tmp);
             NpyIter_Deallocate(iter);
             return -1;
         }
+        Py_DECREF(op_tmp);
```
Comment (pv) [inline numpy/core/src/umath/ufunc_object.c]: Now that I look at this again, I guess `op_tmp` should be placed back into `op` instead of being discarded.

File: doc/source/reference/c-api.iterator.rst [hunk 5]
```diff
@@ -591,6 +619,15 @@ Construction and Destruction
             returns true from the corresponding element in the ARRAYMASK
             operand.
 
+        .. c:var:: NPY_ITER_OVERLAP_NOT_SAME
+
+            In the memory overlap checks done when ``NPY_ITER_COPY_IF_OVERLAP``
+            is specified, consider this array as overlapping even if it is
+            exactly the same as another array.
```
Comment (charris) [inline doc/source/reference/c-api.iterator.rst]: ?. What other array? I think the operation here could be best explained with the help of some diagrams showing the data access and writebacks. The operation can be performed, and is valid, even if there is no overlap, just that it isn't necessary if no overlap.

Comment (charris) [inline doc/source/reference/c-api.iterator.rst]: I think this might be help clarify when the indexing is many to one in either array.

File: doc/source/reference/c-api.iterator.rst [hunk 6]
```diff
@@ -461,6 +461,34 @@ Construction and Destruction
...
+            in the number of dimensions, the decision is made based
+            on heuristics, which has false positives (needless copies in unusual
+            cases) but has no false negatives.
+
+            If read/write overlap exists and write operands are modified in the
+            iterator loop element-wise, this flag ensures the result of the
+            operation is the same as if all operands were copied.
+            In cases where copies would need to be made, **the result of the
+            computation may be undefined without this flag!**
+
```
Comment (charris) [inline doc/source/reference/c-api.iterator.rst]: This is only for two operands? For inplace operations, both inputs are read, but only one is written too. Or does this apply for three operands, two inputs and one output. Or is it more general than that?

File: numpy/core/src/multiarray/mapping.c [hunk 7]
```diff
@@ -722,6 +723,38 @@ prepare_index(PyArrayObject *self, PyObject *index,
 
 
 /**
+ * Check if self has memory overlap with one of the index arrays, or with extra_op.
+ *
+ * @returns 1 if memory overlap found, 0 if not.
+ */
+NPY_NO_EXPORT int
+index_has_memory_overlap(PyArrayObject *self,
```
Comment (charris) [inline numpy/core/src/multiarray/mapping.c]: Hmm, so you are actually checking indexes rather than the resulting view?

**Comment by charris** [comment]
Just a first read through. This will need a release note eventually.

**Comment by charris** [comment]
Probably should bring in @seberg on this also.

File: doc/source/reference/c-api.iterator.rst (hunk 6 above)
Comment (pv) [inline doc/source/reference/c-api.iterator.rst]: It applies to any number of operands.

File: numpy/core/src/multiarray/mapping.c (hunk 7 above)
Comment (pv) [inline numpy/core/src/multiarray/mapping.c]: It's used to checking the overlap between the array operated on, and the index array used in `ufunc.at`, for example
```
x = np.arange(100)
np.invert.at(x[::-1], x)  # <- oob memory access without x.copy() of the second argument
```

File: numpy/core/tests/test_mem_overlap.py [hunk 8]
```diff
@@ -518,5 +564,363 @@ def __array__(self):
...
+    interface['descr'] = [('', '|b1')]
+    return np.asarray(DummyArray(interface, x))
+
+
+class TestUFunc(object):
+    """
+    Test ufunc call memory overlap handling
+    """
+
+    def check_unary_fuzz(self, operation, get_out_axis_size, dtype=np.int16,
```
Comment (pv) [inline numpy/core/tests/test_mem_overlap.py]: I would need to check again if these tests cover all places where the copyifoverlap behavior is in, and/or whether some of the tests are unnecessary.

File: numpy/core/src/umath/ufunc_object.c [hunk 9]
```diff
@@ -3635,6 +3654,7 @@ PyUFunc_Reduceat(PyUFuncObject *ufunc, PyArrayObject *arr, PyArrayObject *ind,
         /* In case COPY or UPDATEIFCOPY occurred */
         op[0] = NpyIter_GetOperandArray(iter)[0];
         op[1] = NpyIter_GetOperandArray(iter)[1];
+        op[2] = NpyIter_GetOperandArray(iter)[2];
```
Comment (juliantaylor) [inline numpy/core/src/umath/ufunc_object.c]: why is this line added? a bug in the old code or due to the other changes?

Comment (pv) [inline numpy/core/src/umath/ufunc_object.c]: I think it's a noop. Not necessary to add.

Comment (pv) [inline numpy/core/src/umath/ufunc_object.c]: On second thought, I think it's useful to have here to ensure correctness, if we later on change NPY_ITER_COPY_IF_OVERLAP to make copies of also the input arrays.

**Comment by pv** [comment]
Fixed one bug an re-checked test coverage.

Here's a question: what should the following operation produce:
```
>>> import numpy as np
>>> from numpy.lib.stride_tricks import as_strided
>>> x = as_strided(np.array([1]), (2,), (0,))
>>> np.add(x, x.copy(), x)
# array([3, 3])   OR
# array([2, 2]), as in np.add(x.copy(), x.copy(), x)
```
There are some tests in `test_nditer.py` that expect that `y[...] += x` gives the [3,3] type result when `y` has zero strides. Can we chalk this off as relying on undefined behavior?

**Comment by seberg** [comment]
It is a problem, short of special casing these some kind of overlaps as being OK, there is nothing we can do. But, I do expect I even have some very old dirty scripts that might break by this, which doesn't mean much, just saying that I do think that some breakage will probably occur in some weirder scripts.

However, the only possible way to move forward I see is to first of all hope that nothing really breaks, and if it does (or do we have to assume it will?), it would be starting off with a FutureWarning and doing the actual change later. I guess in principle the futurewarning could be given only in cases that were not complete bogus before/easy to abuse, which probably is limited to identical strides.

**Comment by pv** [comment]
I think the earlier checks showed that mainstream projects such as sklearn, pandas, scipy etc. do not really have overlapping arrays in their testsuites. So I think we may need to make the decision on what is a FutureWarning, without knowing example use in the wild.

What should be FutureWarning --- everything, or just ovelapping-output-has-internal-overlap case?
If we want to be conservative, we could just make this just emit futurewarnings first, and then turn it on for reals only in the next release. Or maybe also make it conditional on some compile-time flag such as the relaxed strides stuff was.

**Comment by seberg** [comment]
Its cool this is getting close. A compile time flag may be an option. Overall, I don't think any major project is likely to run into this, and any project with a test suit should notice immidiately. I am a bit scared of breaking stuff, but on the other hand, I don't really believe we will break a lot (or if, the break is probably visible at least). To be honest, I think we probably had this discussion before, and the examples I thought of last time were incorrect ;).

Frankly, maybe I am optimistic, but I think we may just want to give it a shot....

**Comment by mhvk** [comment]
I'm coming very late to this, but I'm worried about this test case:
```
       x = np.add(a, a, a)
       assert_(not np.shares_memory(x, a))  # returned array is always new
```
Previous it was guaranteed that if you passed in an `out` argument, that would be the one returned (indeed, there is currently #8416 to remind us that this is not the case for `MaskedArray`). Is this change really wanted? I thought the point was to make temporary copies, which are then stored back in the output array if needed. Does this change imply that `a += a` also returns a new array?

This example seems particularly strange, since it is one where overlap really isn't a problem and thus is more likely to be seen in the wild...

**Comment by pv** [comment]
@mhvk: I don't think there are changes vs `__array_prepare__` behavior. (I tested vs astropy and all tests pass.) The comment "returned array is always new" is due to the specific `__array_prepare__` method always returning a new output array no matter what.

**Comment by pv** [comment]
I'm actually not sure if `__array_prepare__` returning a fully new output array is fully indented behavior, as it's apparently not allowed with `where=`, see [current code in master branch](https://github.com/numpy/numpy/blob/e0cf8393d3b7e8bb6ae9f7c9871bf0a90e3b4896/numpy/core/src/umath/ufunc_object.c#L1799). It also appears gufuncs don't support array_prepare at all? All this is not really related to this PR, but this stuff looks a bit half-cooked (and I guess it will be obsoleted by numpy_ufunc_override, so no need to fix it...).

**Comment by mhvk** [comment]
OK, I indeed also checked astropy with your PR, and things pass, plus the test case I had in mind. Sorry for not realising the test was more unusual in calling `np.array`... I actually don't think one is supposed to be able to return anything but a view in `__array_prepare__`...

**Comment by mhvk** [comment]
From the documentation [1], it seems indeed that `__array_prepare__` should not touch/copy the output data (though another page [2] is a bit less explicit):
"In addition to __array_wrap__, which is called on the way out of the ufunc, there is also an __array_prepare__ method which is called on the way into the ufunc, after the output arrays are created but before any computation has been performed. The default implementation does nothing but pass through the array. __array_prepare__ should not attempt to access the array data or resize the array, it is intended for setting the output array type, updating attributes and metadata, and performing any checks based on the input that may be desired before computation begins. Like __array_wrap__, __array_prepare__ must return an ndarray or subclass thereof or raise an error."

[1] https://docs.scipy.org/doc/numpy/user/basics.subclassing.html
[2] https://docs.scipy.org/doc/numpy/reference/arrays.classes.html

**Comment by njsmith** [comment]
Historically, we've said that stuff like `np.add(a, b, a)` where the overlap is confined to within each "core" operation is well-defined, and that everything else involving overlap between inputs and outputs, or between outputs and outputs (like the 0-strided output case), is undefined. Plus we've regularly broken the overlap cases over the last few years (e.g. due to implementing iteration order optimizations), which has caused a few people to complain but wasn't a disaster.

Doc issue: the current PR doesn't make clear what happens to the cases that were always well-defined (`np.add(a, b, a)` etc.). I assume that they continue to be well-defined, and that they continue *not* to make a copy? This should be explicit though - if you take the current changes.rst text literally then I think it implies that they now make an (unnecessary) copy. I'm also not 100% clear on what this patch does for weirdnesses like `np.add(a, b, out=np.as_strided(np.array([1]), shape=(2,), strides=(0,))` -- I assume this still invokes undefined behavior? (I guess the ideal behavior here would just be an error, b/c what is that even supposed to mean.) It'd be nice if there were a bit of text in the docs that lays out all the different cases and how they're handled.

For the example
```python
>>> x = as_strided(np.array([1]), shape=(2,), strides=(0,))
>>> np.add(x, x.copy(), x)
```
then `[3, 3]` seems reasonable to me, if we even want to support it at all. (This is only well-defined because it happens that all the outputs are identical, right? `np.add(x, [1, 2], x)` is ill-defined regardless of this patch, because even `np.add([1, 1], [1, 2], x)` is already ill-defined.) I'm sorta feeling like we don't need any `FutureWarning`s here, because the "breakage" is no more than regularly happens due to like, memory layout changes. OTOH if we wanted to be conservative and issue `FutureWarning`s then that would be OK with me too, and I'd vote that instead of trying to be super-clever about detecting which cases to issue the warnings in, we just run the new logic for deciding when to copy and have a release or two that issues a `FutureWarning` in *every* case where the make-a-copy-logic activates.

**Comment by njsmith** [comment]
(Also, this is really awesome and exciting!)

**Comment by pv** [comment]
The change here is that all ufuncs operations produce the same results as similar operation on arrays that have the same content but no read/write memory overlap. For overlapping arrays, copies are skipped if the data dependency is simple enough --- currently, this is only done in the "exactly same array" and "trivial 1d iteration with simple data dependence" cases. So yes, `np.add(a, b, a)` will not make a copy (unless `a` happens to have internal overlap in which case a copy is made).

I think the general case of zero-stride output that overlaps with inputs is undefined. The `[3,3]` indeed appears sort of reasonable because the arrays are the same. However, this would be an ugly corner case if it is left in IMHO (so it's currently removed), as it breaks the "`x += y` is the same as `x[...] = x + y`" rule.

Although we *could* also detect and handle the case where the only overlap is internal overlap in the write operand, I didn't try to change the behavior here, it is still undefined.

**Comment by njsmith** [comment]
Oh, sorry, I misread on the `[3, 3]` thing - I meant that I don't see any problem with changing it, but got mixed up and thought `[3, 3]` was the "new" output. I should have said `[2, 2]` :-)

**Comment by njsmith** [comment]
Okay, so to rephrase for my own understanding:

* It used to be that ufunc operations were ill-defined when there was overlap between one of the input arrays and one of the output arrays, *except* in the case where the overlap was only between scalars at identical indices.

* Now, ufunc operations are well-defined regardless of overlap between inputs and outputs. An operation `foo(a, b, out=(b, c))` is defined to be equivalent to the same operation where all the inputs are replaced by copies of them. However, in some cases we are able to optimize out these copies (this includes both the cases that were previously well-defined, but also some new cases like your example `np.add(x[1:], x[:-1], x[:-1])` that used to be ill-defined!). One can imagine further optimizations like this in the future, but since the semantics are now well-defined this would be purely an internal thing.

* Everything above only discussed overlap between an input array and an output array. There are two more cases to mention for completeness. For overlap between an input array and an input array (possibly itself), there's never been any problem and there still isn't -- `add(a, a.T)` or `add(zero_strided_a, b)` have always been well-defined and still are. For overlap between an output array and an output array (like `out=(a, a)`, or `out=zero_strided_a`, or ...), then it was always ill-defined, and remains ill-defined.

That all makes sense and sounds good to me! But...

> (unless `a` happens to have internal overlap in which case a copy is made)

...this line confuses me then :-). According to my analysis above, internal overlap shouldn't affect anything -- internal overlap in an input array doesn't matter, and internal overlap in an output array makes the whole operation ill-defined. So I must be missing something. When does internal overlap trigger a copy and why?

File: numpy/core/src/private/lowlevel_strided_loops.h (hunk 1 above)
Comment (pv) [inline numpy/core/src/private/lowlevel_strided_loops.h]: OTOH, I don't really want to put it in a .c file since the logic is tightly coupled to the macros here. Removing the inline makes it emit compiler warnings. I could write this as a macro, but that's messy.

File: numpy/core/tests/test_mem_overlap.py (hunk 8 above)
Comment (pv) [inline numpy/core/tests/test_mem_overlap.py]: ok, checked --- removing any of `NPY_COPY_IF_OVERLAP` et al. causes some test to fail

**Comment by pv** [comment]
@njsmith: it's internal overlap in an output array that also overlaps with an input. I guess handling this case wouldn't be strictly necessary if we declare it an undefined operation. It just happens that numpy test suite had two cases of this (trying to be clever and writing reductions using `+=` --- but this being unrelated to what was being tested).

**Comment by njsmith** [comment]
It seems to me that those tests are just wrong though then... I mean, we can make the copy, it doesn't matter much either way I guess, but either way the final result is still going to be dependent on the internal ufunc iteration order, right?

**Comment by pv** [comment]
It's order-independent (but probably not buffering independent), as it's
a simple scalar reduction case.

**Comment by njsmith** [comment]
What I mean is, if we have `annoying = np.as_strided(np.array([1]), shape=(2,), strides=(0,)`, then `np.add([1, 1], [1, 2], out=annoying)` could reasonably give either `[2, 2]` or `[3, 3]`, depending on whether it writes the output array from front-to-back or back-to-front. We're writing both 2 and 3 to the same location, but we don't guarantee which will win.

**Comment by pv** [comment]
That's correct, but I think the specific reduction case in the tests I
wrote about was order-independent as it's `np.add([1, 2], annoying,
out=annoying)`. In any case, those tests are now rewritten, so if the
understanding is that this is in general undefined behavior, there are
no further changes required.

**Comment by njsmith** [comment]
Oh, right, now I understand! ...this rings a bell, too. @seberg, weren't you telling me at some point about how there's some code inside numpy that relies on implementing reducations via zero-strided calls to the non-reduce loops? Would that be affected by this? (My guess is that whatever code is using this trick is "below" the overlap checking in the stack so unaffected by it, and I'm totally happy with this not being something we guarantee to end-users. Also hopefully no-one is writing horrible `as_strided` expressions as a "shorthand" instead of calling `np.sum`!)

**Comment by pv** [comment]
The copies are triggered only if the iterator is given 
NPY_ITER_COPY_IF_OVERLAP flag, which I was careful to set only if I 
understand the code --- so it is not set for the reduction iterators, 
and the overlap issue is handled before the iterator.

**Comment by charris** [comment]
@pv Do you feel that this is about ready?

**Comment by pv** [comment]
@charris: I've read this several times through now, and only finding forgotten decref on error handling. I'm fairly confident now that it's what it says on the tin, and doesn't do anything more.

It's a bit ugly that this adds `PyArray_MapIterArrayCopyIfOverlap` to the API --- it's essentially a private function, as was `PyArray_MapIterArray` --- but the better way is not obvious. On the other hand, adding the new iterator flags appears to be a good way to do this, as it makes reasoning about the change easier, and the changes in ufuncs themselves are mostly minimal.

**Comment by pv** [comment]
One additional thing is that this probably won't work on PyPy due to the use of updateifcopy, but I guess that's part of a bigger pypy issue that needs separate fixes (probably sprinkling `PyArray_FlushUpdateIfCopy` around) in more places than just this.

**Comment by charris** [comment]
@mattip @rlamy Comments?

**Comment by rlamy** [comment]
Well, I don't really understand the UPDATEIFCOPY issue, but since @mattip is on holiday, I'll comment anyway.

All the new tests appear to pass on both pypy and pypy3. I'm guessing that's because the copy cannot be seen by the RPython GC. So I think this is OK for us.

**Comment by pv** [comment]
Yes, the updateifcopy arrays here are never passed back to Python. If 
cpyext is effectively refcounted in this case, then there should be no 
problem. I guess things would break also in other places if this wasn't 
the case, as the iterator uses updateifcopy temporaries also for other 
purposes.

File: numpy/core/src/multiarray/mapping.c [hunk 10]
```diff
@@ -704,6 +705,38 @@ prepare_index(PyArrayObject *self, PyObject *index,
...
+                         int index_type, npy_index_info *indices, int num,
+                         PyObject *extra_op)
+{
+    int i;
+
+    if (index_type & (HAS_FANCY | HAS_BOOL)) {
+        for (i = 0; i < num; ++i) {
+            if (indices[i].object != NULL && PyArray_Check(indices[i].object) &&
+                solve_may_share_memory(self, (PyArrayObject *)indices[i].object,
```
Comment (charris) [inline numpy/core/src/multiarray/mapping.c]: Should have extra indent here to separate the condition from the block.

Comment (charris) [inline numpy/core/src/multiarray/mapping.c]: Bit tricky with the long line, but it doesn't read well as is.

File: numpy/core/src/multiarray/mapping.c [hunk 11]
```diff
@@ -704,6 +705,38 @@ prepare_index(PyArrayObject *self, PyObject *index,
...
+                                       1) != 0) {
+
+                return 1;
+            }
+        }
+    }
+
+    if (extra_op != NULL && PyArray_Check(extra_op) &&
+        solve_may_share_memory(self, (PyArrayObject *)extra_op, 1) != 0) {
```
Comment (charris) [inline numpy/core/src/multiarray/mapping.c]: See above.

File: numpy/core/src/multiarray/mapping.c [hunk 12]
```diff
@@ -3197,6 +3259,21 @@ PyArray_MapIterArray(PyArrayObject * a, PyObject * index)
 }
 
 
+/*NUMPY_API
+ *
+ * Use advanced indexing to iterate an array. Please note
```
Comment (charris) [inline numpy/core/src/multiarray/mapping.c]: This needs to be documented somewhere public. It seems dangerous to expose an API that may change...

Comment (charris) [inline numpy/core/src/multiarray/mapping.c]: How could someone detect the change to this particular function?

File: numpy/core/src/multiarray/nditer_constr.c [hunk 13]
```diff
@@ -2708,6 +2710,80 @@ npyiter_allocate_arrays(NpyIter *iter,
         bufferdata = NIT_BUFFERDATA(iter);
     }
 
+    if (flags & NPY_ITER_COPY_IF_OVERLAP) {
+        /* Perform operand memory overlap checks if requested */
```
Comment (charris) [inline numpy/core/src/multiarray/nditer_constr.c]: Could add more detail on what is being checked here, the notes below are a bit too local.

File: numpy/core/src/multiarray/nditer_constr.c [hunk 14]
```diff
@@ -2708,6 +2710,80 @@ npyiter_allocate_arrays(NpyIter *iter,
...
+
+                    continue;
+                }
+
+                /*
+                 * Use max work = 1. If the arrays are large, it might
+                 * make sense to go further.
+                 */
+                may_share_memory = solve_may_share_memory(
+                    op[iop], op[iother], 1);
```
Comment (charris) [inline numpy/core/src/multiarray/nditer_constr.c]: Maybe extra indent here, or even break the args and align.

File: numpy/core/src/private/lowlevel_strided_loops.h (hunk 1 above)
Comment (charris) [inline numpy/core/src/private/lowlevel_strided_loops.h]: C++ beckons ;)

File: numpy/core/src/private/lowlevel_strided_loops.h [hunk 15]
```diff
@@ -673,6 +691,63 @@ npy_bswap8_unaligned(char * x)
...
+
+    if (arr1_read && arr2_read) {
+        return 1;
+    }
+
+    if (solve_may_share_memory(arr1, arr2, 1) == 0) {
+        return 1;
+    }
+
+    /* Arrays overlapping in memory may be equivalently iterable
```
Comment (charris) [inline numpy/core/src/private/lowlevel_strided_loops.h]: Multiline comment not correctly formatted ;) Also below.

File: numpy/core/src/private/lowlevel_strided_loops.h [hunk 16]
```diff
@@ -673,6 +691,63 @@ npy_bswap8_unaligned(char * x)
...
+    /* Arrays with zero stride are never "ahead" since the element is reused
+       (at this point we know the arrays do overlap). */
+
+    if (stride1 > 0) {
+        arr1_ahead = (stride1 >= stride2 &&
+                      (npy_uintp)PyArray_BYTES(arr1) >= (npy_uintp)PyArray_BYTES(arr2));
+    }
+    else if (stride1 < 0) {
+        arr1_ahead = (stride1 <= stride2 &&
+                      (npy_uintp)PyArray_BYTES(arr1) <= (npy_uintp)PyArray_BYTES(arr2));
```
Comment (charris) [inline numpy/core/src/private/lowlevel_strided_loops.h]: Hmm... Out of curiousity, did the compiler complain without the cast?

File: numpy/core/src/umath/ufunc_object.c [hunk 17]
```diff
@@ -1730,12 +1764,17 @@ execute_fancy_ufunc_loop(PyUFuncObject *ufunc,
         }
     }
     for (i = nin; i < nop; ++i) {
+        /*
+         * Because we don't write to all elements, the output arrays
+         * must be considered READWRITE by the iterator.
```
Comment (charris) [inline numpy/core/src/umath/ufunc_object.c]: Curious.

**Comment by charris** [comment]
Not done reviewing yet. Determining when a copy isn't needed for overlapping arrays seems tricky except in single operand the linear case. ISTR that the iterator may coalesce axes.

File: numpy/core/src/multiarray/mapping.c (hunk 12 above)
Comment (pv) [inline numpy/core/src/multiarray/mapping.c]: This the is old comment for `PyArray_MapIterArray`, not added by me.

File: numpy/core/src/umath/ufunc_object.c (hunk 17 above)
Comment (pv) [inline numpy/core/src/umath/ufunc_object.c]: Updateifcopy copies all elements back, not only those for which wheremask is true, so the old content of the output array must be copied there.

File: numpy/core/src/private/lowlevel_strided_loops.h (hunk 16 above)
Comment (pv) [inline numpy/core/src/private/lowlevel_strided_loops.h]: The casts are not needed (probably leftovers from earlier versions).

File: numpy/core/src/multiarray/mapping.c (hunk 12 above)
Comment (pv) [inline numpy/core/src/multiarray/mapping.c]: Anyway, removing it, since it's unlikely we will change this...

**Comment by charris** [comment]
I'm going to put this in so it will start getting more testing. I don't claim to have understood all the details, but I doubt I will discover any errors by inspection. The write up inth release notes is very nice, you should consider copying that over to the official documentation. There are still some style nits that I may attempt to clean up later, I would gently urge you to pay more attention to spaces around operators ('+') and line length, even though the latter may not always be as readable. 

Thanks for this work, Pauli.

**Comment by seberg** [comment]
Wow, cool. Assuming nothing goes haywire, this is the next huge addition :)!
